src/cursor/cur_std.c

```diff
@@ -227,6 +227,49 @@ __wt_cursor_kv_not_set(WT_CURSOR *cursor, bool key) WT_GCC_FUNC_ATTRIBUTE((cold)
       key ? "key" : "value");
 }
 
+/*
+ * __wt_cursor_copy_release_item --
+ *     Release memory used by the key or value item in cursor copy debug mode.
+ */
+int
+__wt_cursor_copy_release_item(WT_CURSOR *cursor, WT_ITEM *item) WT_GCC_FUNC_ATTRIBUTE((cold))
+{
+    WT_DECL_ITEM(tmp);
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+    void *extra_mem;
+
+    session = (WT_SESSION_IMPL *)cursor->session;
+
+    /*
+     * If we own the memory for the item, make a copy and use that instead. That allows us to
+     * overwrite and free the original memory, potentially uncovering programming errors related to
+     * retaining pointers to key/value memory beyond API boundaries. If the item has memory not
+     * currently in use, free that as well.
+     */
+    if (WT_DATA_IN_ITEM(item)) {
+        /*
+         * We need an extra allocation and a free of it at the right time. Otherwise, with some
+         * allocators, we may get the exact same memory address back in the item. and we will never
+         * catch any accesses to the original memory. Ideally, We want the memory originally used by
+         * the item to end up on the malloc free list when we return.
+         */
+        WT_ERR(__wt_malloc(session, item->size, &extra_mem));
+        WT_ERR(__wt_scr_alloc(session, 0, &tmp));
+        WT_ERR(__wt_buf_set(session, tmp, item->data, item->size));
+        __wt_explicit_overwrite(item->mem, item->memsize);
+        __wt_buf_free(session, item);
+        __wt_free(session, extra_mem);
+        WT_ERR(__wt_buf_set(session, item, tmp->data, tmp->size));
+    } else if (item->mem != NULL) {
+        __wt_explicit_overwrite(item->mem, item->memsize);
+        __wt_buf_free(session, item);
+    }
```

Comment:
OK, that took a minute... 🥇 

I worry about gaming the system (for all we know, there's an allocator out there that does FIFO, and with 2 elements on the queue, we'll get the one we don't want).

Do the sanitizers actually not catch access with an intervening free/malloc pair? It would seem to me they **must**, otherwise they'd be useless? Did you actually see this in practice in a sanitizer run?

Anyway, much as it pains me to say it, I'd suggest creating a buffer swap function and explicitly swap the newly allocated `WT_ITEM` for the `WT_ITEM` we don't want.

Reply:
> Do the sanitizers actually not catch access with an intervening free/malloc pair? 
They do catch it.  I was thinking (hoping) to have the option have some good effect even when not using a sanitizer.  I'm realizing it leads to something that is too fragile.  So I think I'll take out the extra allocation, and make the code simpler.  And then it really has to be used with a sanitizer.

Reply:
I'm looking into the buffer swap function.

Reply:
I'm not really loving the buffer swap function idea.  If the buffer swap function does all this stuff, including freeing the old memory, then it will need to do overwrite the old memory (which needs to be done whether or not it is a diagnostic build).  That makes the buffer swap function kind of special case.  It's going to be either ugly here or ugly in `include/buf.i` .  I'll put it in if you still like it, but it's really only going to have a specific use.

Reply:
FYI, the sanitizer use case is the one MongoDB is looking for. So "it really has to be used with a sanitizer" is fine.

Reply:
I agree with @sueloverso, this feature is intended for sanitizers, I'm happy to go with that focus.